### PR TITLE
[fix bug 1363304] Lazy load Firefox feature thumbnail images

### PR DESCRIPTION
**Pull request by alexgibson**
## Description
- Lazy loads /firefox/features/ thumbnails using [IntersectionObserver API](https://developer.mozilla.org/docs/Web/API/Intersection_Observer_API).
- Non supporting browsers load images normally (we could fallback to some JS library that does lazy load, but for now I think this is a good enough improvement).

## Bugzilla link
https://bugzilla.mozilla.org/show_bug.cgi?id=1363304

## Testing
- Demo: https://bedrock-demo-agibson.us-west.moz.works/en-US/firefox/features/
- Intersection Observer doesn't ship in Firefox until 54, so you'll need to test using Beta/Nightly for now.
- Also test on Chrome and Edge which both support Intersection Observer.
- Test noscript fallback.

## Checklist
- [ ] Requires l10n changes.
- [ ] Related functional & integration tests passing.

### CODE REVIEW

**Comment by alexgibson** [comment]
@jpetto would be good to get your thoughts here in terms of approach. Not urgent, just when you have some free time.

**Comment by alexgibson** [comment]
@jpetto we have some more feature pages coming down the pipe currently, so would be good to try and get this in sometime soon?

File: media/js/firefox/features-landing.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+(function() {
+    'use strict';
+
+    var supportsInsersectionObserver = typeof IntersectionObserver !== 'undefined';
+
+    function callback(changes, observer) {
```
Comment (jpetto) [inline media/js/firefox/features-landing.js]: Nit - would be nice for this function to have a slightly less generic name. Maybe `observerCallback`?

File: media/js/firefox/features-landing.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        changes.forEach(function(change) {
+            if (change.intersectionRatio > 0) {
+                change.target.src = change.target.dataset.src;
+                change.target.onload = function() {
```
Comment (jpetto) [inline media/js/firefox/features-landing.js]: Do we need to wait for the image to load before removing the `data-src` attribute?

File: media/css/firefox/features/common.scss [hunk 3]
```diff
@@ -158,10 +160,32 @@ main {
...
         margin-bottom: 20px;
         max-width: 420px;
+        opacity: 1;
+        transition: opacity 0.3s;
         width: 100%;
     }
+
+    img[data-src] {
+    	opacity: 0;
+        display: none;
```
Comment (jpetto) [inline media/css/firefox/features/common.scss]: There is no smaller nit than calling out an extra space.

File: bedrock/firefox/templates/firefox/features/index.html [hunk 4]
```diff
@@ -203,7 +242,10 @@ <h2 class="section-title"><span>{{ _('Independence features') }}</span></h2>
       </li>
       <li class="features-list-item">
         <a href="{{ url('firefox.features.independent') }}" data-link-type="link" data-link-name="Protect Your Rights" data-link-position="independence 3">
-          <img src="{{ static('img/firefox/features/thumbnails/independent.jpg') }}" alt="">
+          <img src="{{ static('img/firefox/features/thumbnails/placeholder.png') }}" data-src="{{ static('img/firefox/features/thumbnails/independent.jpg') }}" alt="">
+          <noscript>
+            <img src="{{ static('img/firefox/features/thumbnails/independent.jpg') }}" alt="">
+          </noscript>
```
Comment (jpetto) [inline bedrock/firefox/templates/firefox/features/index.html]: This is getting close to a macro...

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/features/index.html]: Hmm, I like this suggestion 👍

File: media/js/firefox/features-landing.js (hunk 2 above)
Comment (alexgibson) [inline media/js/firefox/features-landing.js]: This is just for the little fade-in, seems a bit nicer than just swapping out the image as it plops in.

File: media/js/firefox/features-landing.js (hunk 1 above)
Comment (alexgibson) [inline media/js/firefox/features-landing.js]: Good suggestion

File: media/js/firefox/features-landing.js (hunk 2 above)
Comment (jpetto) [inline media/js/firefox/features-landing.js]: Ah, yeah, agreed.

**Comment by jpetto** [comment]
Btw, tested in Fx Nightly, Safari, Chrome, Edge, IE 11, and IE 9 and all did the right thing.

**Comment by alexgibson** [comment]
@jpetto added a commit with some fixes. Thanks for suggesting the macro 👍

**Comment by alexgibson** [comment]
Squashed :shipit:
